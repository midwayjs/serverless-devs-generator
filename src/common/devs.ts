import { BaseGenerator } from './base';
import { existsSync, readFileSync } from 'fs';
import { join } from 'path';
import { GenerateOptions } from '../interface';
const ymlName = ['s.yaml', 's.yml'];

export abstract class ServerlessDevsGenerator<
  FunctionConfig = unknown
> extends BaseGenerator {
  static canSupport(options: GenerateOptions) {
    // check yaml file exists
    for (const name of ymlName) {
      const filePath = join(options.appDir, name);
      if (existsSync(filePath)) {
        const file = readFileSync(filePath, 'utf8');
        if (this.isPlatformSupport(file)) {
          return [filePath, file];
        }
        return false;
      }
    }
    return false;
  }

  static isPlatformSupport(file: string) {
    return false;
  }
}
