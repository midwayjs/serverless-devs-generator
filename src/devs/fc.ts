import { Document, visit, Scalar, YAMLMap } from 'yaml';
import { RouterInfo } from '@midwayjs/core';
import { ServerlessDevsGenerator } from '../common/devs';

type FunctionConfig = {
  function: {
    name?: string;
    handler?: string;
  };
  triggers?: Array<{
    name: string;
    type:
      | 'http'
      | 'oss'
      | 'timer'
      | 'log'
      | 'mns_topic'
      | 'cdn_events'
      | 'tablestore'
      | 'eventbridge'
      | string;
    config: any;
  }>;
};

export class FcGenerator extends ServerlessDevsGenerator<FunctionConfig> {
  analyzeFunction(result: RouterInfo[]) {
    const allFunc: {
      [functionName: string]: FunctionConfig;
    } = {};
    if (Array.isArray(result)) {
      for (const func of result) {
        if (!func.functionTriggerName) {
          if (func.fullUrl && func.requestMethod && func.funcHandlerName) {
            func.functionTriggerName = 'http';
            func.functionTriggerMetadata = {
              method: func.requestMethod,
              path: func.fullUrl,
            };
          } else {
            continue;
          }
        }
        const handler = func.funcHandlerName;
        if (!handler || func.functionName?.includes('undefined')) {
          continue;
        }

        if (!func.functionTriggerMetadata) {
          func.functionTriggerMetadata = {};
        }

        const funcName =
          func.functionMetadata?.functionName ||
          func.functionTriggerMetadata?.functionName ||
          func.functionName ||
          handler.replace(/[^\w]/g, '-');

        if (!allFunc[funcName]) {
          allFunc[funcName] = {
            function: {
              name: funcName,
              handler,
            },
            triggers: [],
          };
        }

        if (!allFunc[funcName]?.triggers) {
          allFunc[funcName].triggers = [];
        }

        if (!allFunc[funcName]?.function.handler) {
          allFunc[funcName].function.handler = handler;
        }

        delete func.functionTriggerMetadata.functionName;
        delete func.functionTriggerMetadata.middleware;

        const triggerType = func.functionTriggerName;
        let triggerName = func.functionTriggerMetadata?.name;

        if (!triggerName) {
          triggerName =
            triggerType + '-' + this.createName(func.functionTriggerMetadata);
        }

        // 发布时候需要触发器名字，只允许第一个http触发器叫 http
        const triggerNameExists = allFunc[funcName].triggers.find(trigger => {
          return trigger.name === triggerName;
        });

        // 如果名字不唯一，则报错
        if (triggerNameExists) {
          throw new Error(
            `Current trigger name "${triggerName}" not unique, please configure name in @ServerlessTrigger.`
          );
        }

        if (triggerType === 'http') {
          // 对 http 方法 any 的支持
          const { method } = func.functionTriggerMetadata;
          let methodList = [].concat(method || []);
          if (methodList.includes('any') || methodList.includes('all')) {
            func.functionTriggerMetadata.method = 'any';
            methodList = ['GET', 'POST', 'PUT', 'DELETE', 'HEAD'];
          }

          func.functionTriggerMetadata.methods = methodList.map(method => {
            return method.toUpperCase();
          });

          func.functionTriggerMetadata.authType =
            func.functionTriggerMetadata.authType || 'anonymous';
        }

        const newTriggerConfig = {};
        for (const key in func.functionTriggerMetadata) {
          if (this.allowTriggerConfig(triggerType).includes(key)) {
            newTriggerConfig[key] = func.functionTriggerMetadata[key];
          }
        }

        allFunc[funcName].triggers.push({
          name: triggerName,
          type: triggerType,
          config: newTriggerConfig,
        });
      }
    }

    return Object.values(allFunc);
  }

  fillYaml(
    document: Document | Document.Parsed<any, true>,
    result: FunctionConfig[]
  ) {
    let originComponentNode;

    visit(document, {
      Value: (key, node: Scalar, path) => {
        if (node.value === 'devsapp/fc') {
          const fcComponentNode = path[path.length - 2] as YAMLMap;
          if (!originComponentNode) {
            originComponentNode = fcComponentNode;
          }
          const fnName = fcComponentNode.getIn(['props', 'function', 'name']);
          // 查找同名的函数节点
          const findFunctionConfigIdx = result.findIndex(functionConfig => {
            return functionConfig.function.name === fnName;
          });

          // 如果找到同名的，则填充函数信息，并且在原数组中移除
          if (findFunctionConfigIdx !== -1) {
            // 将信息填充到 yaml 节点中
            const functionConfig = result[findFunctionConfigIdx];
            this.setNodeObjectValue(
              ['props'],
              functionConfig,
              fcComponentNode,
              ['name']
            );
            // 删除该节点
            result.splice(findFunctionConfigIdx, 1);
          }
        }
      },
    });

    if (!result.length) {
      return document;
    }

    // 填充新的函数信息
    for (const functionConfig of result) {
      // 克隆现有节点，移除函数名等信息
      const newFcComponentNode = originComponentNode.clone();
      newFcComponentNode.deleteIn(['props', 'function']);
      newFcComponentNode.deleteIn(['props', 'triggers']);
      newFcComponentNode.deleteIn(['props', 'customDomains']);

      const newProjectName = this.createName(functionConfig);
      this.setNodeObjectValue(['props'], functionConfig, newFcComponentNode);
      document.addIn(
        ['services', 'project-' + newProjectName],
        newFcComponentNode
      );
    }

    return document;
  }

  allowTriggerConfig(type: string): string[] {
    switch (type) {
      case 'http':
        return ['authType', 'disableURLInternet', 'methods'];
      case 'timer':
        return ['cronExpression', 'enable', 'payload'];
      case 'oss':
        return ['bucketName', 'events', 'filter'];
      case 'log':
        return [
          'logConfig',
          'sourceConfig',
          'jobConfig',
          'functionParameter',
          'enable',
        ];
      case 'mns_topic':
        return [
          'topicName',
          'region',
          'notifyContentFormat',
          'notifyStrategy',
          'filterTag',
        ];
      case 'cdn_events':
        return ['eventName', 'eventVersion', 'notes', 'filter'];
      case 'tablestore':
        return ['instanceName', 'tableName'];
    }
  }

  static isPlatformSupport(file): boolean {
    return /devsapp\/fc/.test(file);
  }
}
