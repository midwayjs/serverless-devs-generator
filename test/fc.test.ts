import { join } from 'path';
import { FcGenerator } from '../src/devs/fc';
import { Document } from 'yaml';

describe('test serverless devs yaml generate fc case', () => {
  it('should test yaml parse and stringify', () => {
    const [file, content] = FcGenerator.canSupport(join(__dirname, './fixtures/base')) as string[];
    expect(file).toEqual(join(__dirname, './fixtures/base/s.yaml'));
    expect(content).toMatchSnapshot();
  });

  describe('analyzeFunction', function() {
    it('should transform single http trigger', function() {
      const generator = new FcGenerator('', '', '');
      const result = generator.analyzeFunction([{
        'prefix': '/',
        'routerName': '',
        'url': '/hello',
        'requestMethod': 'post',
        'method': 'hello',
        'handlerName': 'helloService.hello',
        'funcHandlerName': 'helloService.hello',
        'controllerId': 'helloService',
        'middleware': [],
        'controllerMiddleware': [],
        'requestMetadata': [],
        'responseMetadata': [],
        'functionName': 'helloService-hello',
        'functionTriggerName': 'http',
        'functionTriggerMetadata': {
          'path': '/hello',
          'method': 'post'
        },
        'functionMetadata': {
          'functionName': 'helloService-hello'
        },
        'fullUrl': '/hello',
        'fullUrlFlattenString': '/hello'
      }]);

      expect(result).toMatchSnapshot();
    });

    it('should transform multi http trigger with same function', function() {
      const generator = new FcGenerator('', '', '');
      const result = generator.analyzeFunction([{
        'prefix': '/',
        'routerName': '',
        'url': '/hello',
        'requestMethod': 'post',
        'method': 'hello',
        'handlerName': 'helloService.hello',
        'funcHandlerName': 'helloService.hello',
        'controllerId': 'helloService',
        'middleware': [],
        'controllerMiddleware': [],
        'requestMetadata': [],
        'responseMetadata': [],
        'functionName': 'helloService-hello',
        'functionTriggerName': 'http',
        'functionTriggerMetadata': {
          'path': '/hello',
          'method': 'post'
        },
        'functionMetadata': {
          'functionName': 'helloService-hello'
        },
        'fullUrl': '/hello',
        'fullUrlFlattenString': '/hello'
      },
        {
          'prefix': '/',
          'routerName': '',
          'url': '/*',
          'requestMethod': 'get',
          'method': 'hello',
          'description': '',
          'summary': '',
          'handlerName': 'helloService.hello',
          'funcHandlerName': 'helloService.hello',
          'controllerId': 'helloService',
          'middleware': [],
          'controllerMiddleware': [],
          'requestMetadata': [],
          'responseMetadata': [],
          'functionName': 'helloService-hello',
          'functionTriggerName': 'http',
          'functionTriggerMetadata': {
            'path': '/*',
            'method': 'get'
          },
          'functionMetadata': {
            'functionName': 'helloService-hello'
          },
          'fullUrl': '/*',
          'fullUrlFlattenString': '/(.*)'
        }
      ]);

      expect(result).toMatchSnapshot();
    });

    it('should transform multi trigger mixin', function() {
      const generator = new FcGenerator('', '', '');
      const result = generator.analyzeFunction([{
        'url': '/hello',
        'requestMethod': 'post',
        'method': 'hello',
        'funcHandlerName': 'helloService.hello',
        'functionName': 'helloService-hello',
        'functionTriggerName': 'http',
        'functionTriggerMetadata': {
          'path': '/hello',
          'method': 'post'
        },
        'functionMetadata': {
          'functionName': 'helloService-hello'
        },
      },
        {
          'url': '/*',
          'requestMethod': 'get',
          'method': 'hello',
          'funcHandlerName': 'helloService.hello',
          'functionName': 'helloService-hello',
          'functionTriggerName': 'http',
          'functionTriggerMetadata': {
            'path': '/*',
            'method': 'get'
          },
          'functionMetadata': {
            'functionName': 'helloService-hello'
          },
        },
        {
          "url": "",
          "requestMethod": "",
          "method": "hello",
          "handlerName": "helloService.timer",
          "funcHandlerName": "helloService.timer",
          "functionName": "helloService-hello",
          "functionTriggerName": "timer",
          "functionTriggerMetadata": {
            "cronExpression": "0 0 4 * * *"
          },
          "functionMetadata": {
            "functionName": "helloService-timer"
          },
        }
      ]);

      expect(result).toMatchSnapshot();
    });

    it('should test http trigger method any', function() {
      const generator = new FcGenerator('', '', '');
      const result = generator.analyzeFunction([{
        'url': '/hello',
        'requestMethod': 'post',
        'method': 'hello',
        'funcHandlerName': 'helloService.hello',
        'functionName': 'helloService-hello',
        'functionTriggerName': 'http',
        'functionTriggerMetadata': {
          'path': '/hello',
          'method': ['any']
        },
        'functionMetadata': {
          'functionName': 'helloService-hello'
        },
      }]);

      expect(result).toMatchSnapshot();
    });

    it('should test http trigger method all', function() {
      const generator = new FcGenerator('', '', '');
      const result = generator.analyzeFunction([{
        'url': '/hello',
        'requestMethod': 'post',
        'method': 'hello',
        'funcHandlerName': 'helloService.hello',
        'functionName': 'helloService-hello',
        'functionTriggerName': 'http',
        'functionTriggerMetadata': {
          'path': '/hello',
          'method': 'all'
        },
        'functionMetadata': {
          'functionName': 'helloService-hello'
        },
      }]);

      expect(result).toMatchSnapshot();
    });

    it('should test unique trigger name', function() {
      const generator = new FcGenerator('', '', '');
      expect(() => {
        generator.analyzeFunction([{
          'url': '/hello',
          'requestMethod': 'post',
          'method': 'hello',
          'funcHandlerName': 'helloService.hello',
          'functionName': 'helloService-hello',
          'functionTriggerName': 'http',
          'functionTriggerMetadata': {
            'path': '/hello',
            'method': 'all',
            'name': 'abc'
          },
          'functionMetadata': {
            'functionName': 'helloService-hello'
          },
        },
          {
            'url': '/hello',
            'requestMethod': 'post',
            'method': 'hello',
            'funcHandlerName': 'helloService.hello',
            'functionName': 'helloService-hello',
            'functionTriggerName': 'http',
            'functionTriggerMetadata': {
              'path': '/bbb',
              'method': 'post',
              'name': 'abc'
            },
            'functionMetadata': {
              'functionName': 'helloService-hello'
            },
          }]);
      }).toThrowError();
    });
  });

  describe('fillYaml', function() {
    it('should fill yaml with single function mixin', function() {
      const generator = new FcGenerator('', '', '');
      const document = generator.fillYaml(new Document({
        services: {
          project1: {
            component: 'devsapp/fc',
            props: {
              region: 'cn-hangzhou',
              service: '${vars.service}',
              function: {
                name: 'helloService-hello',
              },

            }
          }
        }
      }), [
        {
          "function": {
            "handler": "helloService.hello",
            "name": "helloService-hello",
          },
          "triggers": [
            {
              "config": {
                "methods": [
                  "GET",
                  "POST",
                  "PUT",
                  "DELETE",
                  "HEAD",
                ],
              },
              "name": "http-38245ab0e0",
              "type": "http",
            },
          ],
        },
      ]);

      expect(document.toString()).toMatchSnapshot();
    });

    it('should fill yaml with new function', function() {
      const generator = new FcGenerator('', '', '');
      const document = generator.fillYaml(new Document({
        services: {
          project1: {
            component: 'devsapp/fc',
            props: {
              region: 'cn-hangzhou',
              service: '${vars.service}',
              function: {
                name: 'helloService-hello',
              },
            }
          }
        }
      }), [
        {
          "function": {
            "handler": "helloService.hello",
            "name": "helloService-hello-other",
          },
          "triggers": [
            {
              "config": {
                "methods": [
                  "GET",
                  "POST",
                  "PUT",
                  "DELETE",
                  "HEAD",
                ],
              },
              "name": "http-38245ab0e0",
              "type": "http",
            },
          ],
        },
      ]);

      expect(document.toString()).toMatchSnapshot();
    });
  });

});
