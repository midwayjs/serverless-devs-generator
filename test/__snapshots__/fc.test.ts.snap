// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test serverless devs yaml generate fc case analyzeFunction should test http trigger method all 1`] = `
[
  {
    "function": {
      "handler": "helloService.hello",
      "name": "helloService-hello",
    },
    "triggers": [
      {
        "config": {
          "methods": [
            "GET",
            "POST",
            "PUT",
            "DELETE",
            "HEAD",
          ],
        },
        "name": "http-38245ab0e0",
        "type": "http",
      },
    ],
  },
]
`;

exports[`test serverless devs yaml generate fc case analyzeFunction should test http trigger method any 1`] = `
[
  {
    "function": {
      "handler": "helloService.hello",
      "name": "helloService-hello",
    },
    "triggers": [
      {
        "config": {
          "methods": [
            "GET",
            "POST",
            "PUT",
            "DELETE",
            "HEAD",
          ],
        },
        "name": "http-7070135c53",
        "type": "http",
      },
    ],
  },
]
`;

exports[`test serverless devs yaml generate fc case analyzeFunction should transform multi http trigger with same function 1`] = `
[
  {
    "function": {
      "handler": "helloService.hello",
      "name": "helloService-hello",
    },
    "triggers": [
      {
        "config": {
          "methods": [
            "POST",
          ],
        },
        "name": "http-d9892a5b83",
        "type": "http",
      },
      {
        "config": {
          "methods": [
            "GET",
          ],
        },
        "name": "http-d5835c7a08",
        "type": "http",
      },
    ],
  },
]
`;

exports[`test serverless devs yaml generate fc case analyzeFunction should transform multi trigger mixin 1`] = `
[
  {
    "function": {
      "handler": "helloService.hello",
      "name": "helloService-hello",
    },
    "triggers": [
      {
        "config": {
          "methods": [
            "POST",
          ],
        },
        "name": "http-d9892a5b83",
        "type": "http",
      },
      {
        "config": {
          "methods": [
            "GET",
          ],
        },
        "name": "http-d5835c7a08",
        "type": "http",
      },
    ],
  },
  {
    "function": {
      "handler": "helloService.timer",
      "name": "helloService-timer",
    },
    "triggers": [
      {
        "config": {
          "cronExpression": "0 0 4 * * *",
        },
        "name": "timer-5026b54af0",
        "type": "timer",
      },
    ],
  },
]
`;

exports[`test serverless devs yaml generate fc case analyzeFunction should transform single http trigger 1`] = `
[
  {
    "function": {
      "handler": "helloService.hello",
      "name": "helloService-hello",
    },
    "triggers": [
      {
        "config": {
          "methods": [
            "POST",
          ],
        },
        "name": "http-d9892a5b83",
        "type": "http",
      },
    ],
  },
]
`;

exports[`test serverless devs yaml generate fc case fillYaml should fill yaml if old trigger exists 1`] = `
"services:
  project1:
    component: devsapp/fc
    props:
      region: cn-hangzhou
      service: \${vars.service}
      function:
        name: helloService-hello
        handler: helloService.hello
      triggers:
        - config:
            methods:
              - GET
              - POST
              - PUT
              - DELETE
              - HEAD
          name: http-38245ab0e0
          type: http
"
`;

exports[`test serverless devs yaml generate fc case fillYaml should fill yaml with new function 1`] = `
"services:
  project1:
    component: devsapp/fc
    props:
      region: cn-hangzhou
      service: \${vars.service}
      function:
        name: helloService-hello
  project-628a08d1e5:
    component: devsapp/fc
    props:
      region: cn-hangzhou
      service: \${vars.service}
      function:
        handler: helloService.hello
        name: helloService-hello-other
      triggers:
        - config:
            methods:
              - GET
              - POST
              - PUT
              - DELETE
              - HEAD
          name: http-38245ab0e0
          type: http
"
`;

exports[`test serverless devs yaml generate fc case fillYaml should fill yaml with single function mixin 1`] = `
"services:
  project1:
    component: devsapp/fc
    props:
      region: cn-hangzhou
      service: \${vars.service}
      function:
        name: helloService-hello
        handler: helloService.hello
      triggers:
        - config:
            methods:
              - GET
              - POST
              - PUT
              - DELETE
              - HEAD
          name: http-38245ab0e0
          type: http
"
`;

exports[`test serverless devs yaml generate fc case should test yaml parse and stringify 1`] = `
"provider:
  platform: aliyun # 平台
  component: 'devsapp/fc' # 组件名称
"
`;
